syncer/client.go

```diff
@@ -0,0 +1,277 @@
+	"github.com/0xPolygon/polygon-edge/syncer/proto"
+	"github.com/0xPolygon/polygon-edge/types"
+	"github.com/hashicorp/go-hclog"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+const (
+	SyncPeersManagerLoggerName = "sync-peers-manager"
+	statusTopicName            = "syncer/status/0.1"
+	defaultTimeoutForStatus    = 10 * time.Second
+)
+
+type syncPeerClient struct {
+	logger     hclog.Logger
+	network    Network
+	blockchain Blockchain
+
+	subscription           blockchain.Subscription
+	topic                  *network.Topic
+	id                     string // node ID
+	peerStatusUpdateCh     chan *NoForkPeer
+	peerConnectionUpdateCh chan *event.PeerEvent
+}
+
+func NewSyncPeerClient(
+	logger hclog.Logger,
+	network Network,
+	blockchain Blockchain,
+) SyncPeerClient {
+	return &syncPeerClient{
+		logger:                 logger.Named(SyncPeersManagerLoggerName),
+		network:                network,
+		blockchain:             blockchain,
+		id:                     network.AddrInfo().ID.String(),
+		peerStatusUpdateCh:     make(chan *NoForkPeer, 1),
+		peerConnectionUpdateCh: make(chan *event.PeerEvent, 1),
+	}
+}
+
+func (m *syncPeerClient) Start() error {
+	go m.startNewBlockProcess()
+	go m.startPeerEventProcess()
+
+	if err := m.startGossip(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (m *syncPeerClient) Close() {
+	if m.subscription != nil {
+		m.subscription.Close()
+
+		m.subscription = nil
+	}
+
+	close(m.peerStatusUpdateCh)
+	close(m.peerConnectionUpdateCh)
+}
+
+func (m *syncPeerClient) GetPeerStatus(peerID peer.ID) (*NoForkPeer, error) {
+	clt, err := m.newSyncPeerClient(peerID)
+	if err != nil {
+		return nil, err
+	}
+
+	timeoutCtx, cancel := context.WithTimeout(context.Background(), defaultTimeoutForStatus)
+	defer cancel()
+
+	status, err := clt.GetStatus(timeoutCtx, &emptypb.Empty{})
+	if err != nil {
+		return nil, err
+	}
+
+	return &NoForkPeer{
+		ID:       peerID,
+		Number:   status.Number,
+		Distance: m.network.GetPeerDistance(peerID),
+	}, nil
+}
+
+func (m *syncPeerClient) GetConnectedPeerStatuses() []*NoForkPeer {
+	var (
+		ps            = m.network.Peers()
+		syncPeers     = make([]*NoForkPeer, 0, len(ps))
+		syncPeersLock = sync.Mutex{}
+		wg            sync.WaitGroup
+	)
+
+	for _, p := range ps {
+		p := p
+
+		wg.Add(1)
+
+		go func() {
+			defer wg.Done()
+
+			peerID := p.Info.ID
+
+			status, err := m.GetPeerStatus(peerID)
+			if err != nil {
+				m.logger.Warn("failed to get status from a peer, skip", "id", peerID, "err", err)
+			}
+
+			syncPeersLock.Lock()
+
+			syncPeers = append(syncPeers, status)
+
+			syncPeersLock.Unlock()
+		}()
+	}
+
+	wg.Wait()
+
+	return syncPeers
+}
+
+func (m *syncPeerClient) GetPeerStatusUpdateCh() <-chan *NoForkPeer {
+	return m.peerStatusUpdateCh
+}
+
+func (m *syncPeerClient) GetPeerConnectionUpdateEventCh() <-chan *event.PeerEvent {
+	return m.peerConnectionUpdateCh
+}
```

Comment:
What do you think about naming this `GetPeerEventCh`?

Reply:
It's little bit difficult. What does `PeerEvent` means because `Peer` here doesn't mean the one in network layer